src/ConfigParser.cc

```diff
@@ -176,6 +176,12 @@ ConfigParser::UnQuote(const char *token, const char **next)
     if (*s != quoteChar && !errorStr) {
         errorStr = "missing quote char at the end of quoted string";
         errorPos = s - 1;
+
+        // Prevent heap-buffer-overflow
+        if (*s == '\0') {
+            debugs(3, DBG_CRITICAL, "FATAL: " << errorStr << ": " << errorPos);
+            self_destruct();
+        }
```

Comment:
The proposed fix contradicts PreviewMode_ handling further below. Instead of dying here, would it be possible to do the following?

1. Keep this "missing quote char" code intact -- undo all current PR changes.
2. Fix the order of conditions in the "We are expecting a separator after quoted string" `if` statement below, so that `s+1` is not dereferenced when errorStr is set. That errorStr variable must be tested first, before interpreting what lies beyond `s`.

There are more problems in this terrible code, but I hope the above small adjustment will not make it _worse_ while we are waiting for a proper rewrite.

Reply:
> The proposed fix contradicts PreviewMode_ handling further below. Instead of dying here, would it be possible to do the following?
> 
> 1. Keep this "missing quote char" code intact -- undo all current PR changes.
> 2. Fix the order of conditions in the "We are expecting a separator after quoted string" `if` statement below, so that `s+1` is not dereferenced when errorStr is set. That errorStr variable must be tested first, before interpreting what lies beyond `s`.
> 
> There are more problems in this terrible code, but I hope the above small adjustment will not make it _worse_ while we are waiting for a proper rewrite.
Thanks for the review @rousskov!   I have resubmitted